ONSdigital/zebedee
# Datasets review workflow

### What

Add in 'last edited by' information and enforce a workflow. The workflow is simplified, for the sake of getting CMD working end-to-end, but there will still be some rules:

- The user who submitted something for review cannot review their own content changes.
- Once a page has been reviewed it always stays in 'reviewed' even if there are edits to it.
- If a page is saved when 'in progress' or 'awaiting review' then it goes to 'in progress' state and the user who last edited it is stored in the collection JSON data.
- Every save/submit for reviewed/mark as reviewed stores who performed that action in `lastEditedBy` property in the a dataset/version object in the collection JSON.

### How to review

- Checkout `feature/datasets-review-workspace` on Zebedee and Florence.
- Make sure you have a dataset and a new version available to edit. The following steps should work identically with both.

Check that a user can't move their own content changes from the states 'Completed` or `InProgress` to `Reviewed` using the browser console:

1) In the Chrome Devtools console try to update the state of your own in progress or completed page to reviewed. For example
```
fetch("/zebedee/collections/my-collection-name-6ed44c3dbd1304ceb25eb22546c24c5007d5052c1d3cf1e9c0ec5c0682978aea/datasets/mid-year-pop-est/editions/time-series/versions/1", {
	credentials: 'include',
	method: 'PUT',
	body: JSON.stringify({
		state: "Reviewed"		
    })
}).then(response => console.log(response));
```
2) The above code should give a 403 response if you're trying to approve your own changes.
3) Log out and login as a different user. Now attempt to do the same as above in the console but with the different user. It should give a 200 response and the collection JSON data is updated with a 'Reviewed' state for that dataset/version.

### Who can review

Zebedee specifically should have someone like @daiLlew, @CarlHembrough or @ian-kent look at it.
